File: Javascript/Behavioral Patterns/Strategy/strategy.ts
interface PaymentStrategy {
    pay(amount: number): void
}

export class CreditCartPayment implements PaymentStrategy {
    private cardNumber: string

    constructor(cardNumber: string) {
        this.cardNumber = cardNumber
    }

    pay(amount: number): void {
        console.log(`Pay ${amount} with card ${this.cardNumber}`)
    }
}

export class LinePayPayment implements PaymentStrategy {
    private mobile: string

    constructor(mobile: string) {
        this.mobile = mobile
    }

    pay(amount: number): void {
        console.log(`Pay ${amount} with LinePay ${this.mobile}`)
    }
}

export class PaypalPayment implements PaymentStrategy {
    private email: string

    constructor(email: string) {
       this.email = email
    }

    pay(amount: number): void {
        console.log(`Pay ${amount} with PayPal ${this.email}`)
    }
}

export class PaymentContext {
    private strategy: PaymentStrategy
    
    constructor(strategy: PaymentStrategy) {
       this.strategy = strategy 
    }

    setStrategy(strategy: PaymentStrategy) {
        this.strategy = strategy
    }

    executePayment(amount: number): void {
        this.strategy.pay(amount)
    }
}

File: Go/Behavioral_Patterns/Strategy/strategy.go
package strategy

import "fmt"

type PaymentStrategy interface {
	Pay(amount int)
}

type CreditCartPayment struct {
	cardNumber string
}

func (c *CreditCartPayment) Pay(amount int) {
	fmt.Printf("Pay %d with card %s\n", amount, c.cardNumber)
}

func NewCreditCardPayment(cardNumber string) *CreditCartPayment {
	return &CreditCartPayment{
		cardNumber: cardNumber,
	}
}

type PaypalPayment struct {
	email string
}

func (p *PaypalPayment) Pay(amount int) {
	fmt.Printf("Pay %d with paypal %s\n", amount, p.email)
}

func NewPaypalPayment(email string) *PaypalPayment {
	return &PaypalPayment{
		email: email,
	}
}

type LinePayment struct {
	mobile string
}

func (p *LinePayment) Pay(amount int) {
	fmt.Printf("Pay %d with linepay %s\n", amount, p.mobile)
}

func NewLinePayment(mobile string) *LinePayment {
	return &LinePayment{
		mobile: mobile,
	}
}

type PaymentContext struct {
	Strategy PaymentStrategy
}

func (p *PaymentContext) SetPayment(payment PaymentStrategy) {
	p.Strategy = payment
}

func NewPaymentContext(payment PaymentStrategy) *PaymentContext {
	return &PaymentContext{
		Strategy: payment,
	}
}
